general/handle.hpp

```diff
@@ -0,0 +1,200 @@
+// Copyright (c) 2010-2024, Lawrence Livermore National Security, LLC. Produced
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+#ifndef MFEM_HANDLE_HPP
+#define MFEM_HANDLE_HPP
+
+#include "../config/config.hpp"
+#include <memory>
+
+namespace mfem
+{
+
+/// @brief A smart pointer class that may represent either shared ownership, or
+/// a non-owning borrow.
+///
+/// A Handle may either be owning or non-owning. Non-owning Handle%s point to
+/// externally owned data; it is the responsibility of the user both to ensure
+/// that the data remains valid as long as the Handle is alive and to delete the
+/// pointer when it is no longer needed. Owning Handle%s use <a
+/// href="https://en.cppreference.com/w/cpp/memory/shared_ptr">
+/// std::shared_ptr</a> to implement reference counting. The underlying data
+/// will be valid as long as there is at least one live copy. When the last
+/// Handle is destroyed, the pointer is deleted.
+///
+/// Both types of Handle%s can be copied, moved, stored in standard containers,
+/// etc.
+///
+/// A non-owning Handle may assume ownership over its data, but an owning Handle
+/// cannot release ownership over its data.
+///
+/// It is an invariant of this class that **at most** one of the data members
+/// @a not_owned and @a owned will be non-null.
+template <typename T>
+class Handle
+{
+   /// If this is a non-owning handle, @a not_owned will point to the data.
+   T *not_owned = nullptr;
+
+   /// If this is an owning handle, @a owned will point to the data.
+   std::shared_ptr<T> owned = nullptr;
```

Comment:
```suggestion
   std::shared_ptr<T> owned;
```
Initialization to `nullptr` is implicit.